compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+
+    expected_lits = [255, 254, 255, 254, 2, 44]
+    expected_seqs = [
+        LZ77Sequence(0, 4, 3),
+        LZ77Sequence(0, 5, 9),
+        LZ77Sequence(4, 3, 4),
+        LZ77Sequence(1, 6, 22),
+    ]
+    seqs, lits = encoder.lz77_parse_generate_sequences(data_block)
+
+    assert encoder.window == initial_window + data_list
+    assert sum(len(v) for v in encoder.dict.values()) == len(encoder.window) - min_match_len + 1
+    assert lits == expected_lits
+    assert seqs == expected_seqs
+
+    # encode another block which is copy of first and see that we get just one match
+    seqs, lits = encoder.lz77_parse_generate_sequences(data_block)
+    assert encoder.window == initial_window + data_list * 2
+    assert (
+        sum(len(v) for v in encoder.dict.values()) == len(encoder.window) - min_match_len + 1
+    )  # subtract 2 since min_match_len is 3
+    assert lits == []
+    assert seqs == [LZ77Sequence(0, len(data_list), len(data_list))]
+
+    # now reset encoder and verify that after encoding we get results that we expect without initial window
+
+    # matches:
+    # 1, [1, 1, 1,] 0, 0, [1, 1, 1,] 255, 254, 255, 254, 255, 254, 255, 2, 0, 0, 1, 1, 1, 1, 44
+    # 1, 1, 1, 1, 0, 0, 1, 1, 1, [255, 254, 255,] 254, [255, 254, 255,] 2, 0, 0, 1, 1, 1, 1, 44
+    # 1, 1, 1, 1, [0, 0, 1, 1, 1,] 255, 254, 255, 254, 255, 254, 255, 2, [0, 0, 1, 1, 1,] 1, 44
+    encoder.reset()
+    expected_lits = [1, 1, 1, 1, 0, 0, 255, 254, 255, 254, 2, 1, 44]
+    expected_seqs = [LZ77Sequence(6, 3, 5), LZ77Sequence(4, 3, 4), LZ77Sequence(1, 5, 13)]
+    seqs, lits = encoder.lz77_parse_generate_sequences(data_block)
+
+    assert encoder.window == data_list
+    assert sum(len(v) for v in encoder.dict.values()) == len(data_list) - min_match_len + 1
+    assert lits == expected_lits
+    assert seqs == expected_seqs
+
+
+def test_lz77_multiblock_file_encode_decode():
+    """full test for LZ77Encoder and LZ77Decoder
+
+    - create a sample file
+    - encode the file using LZ77Encoder
+    - perform decoding and check if the compression was lossless
+
+    """
+    initial_window = [44, 45, 46] * 5
+    # define encoder, decoder
+    encoder = LZ77Encoder(window_initialization=initial_window)
+    decoder = LZ77Decoder(window_initialization=initial_window)
+
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        # create a file with some random data
+        input_file_path = os.path.join(tmpdirname, "inp_file.txt")
+        create_random_binary_file(
+            input_file_path,
+            file_size=500,
+            prob_dist=ProbabilityDist({44: 0.5, 45: 0.25, 46: 0.2, 255: 0.05}),
+        )
+
+        # test lossless compression
+        assert try_file_lossless_compression(
+            input_file_path, encoder, decoder, encode_block_size=1000
+        )
+
+
+# Provide a simple CLI interface below for convenient experimentation
```

Comment:
define  the parser args after ` if __name__ == "__main__":`
As if we import multiple such files, it will cause issues.

Reply:
done now